File: Answer/3296.移山所需的最少秒数.cs
/*
 * @lc app=leetcode.cn id=3296 lang=csharp
 *
 * [3296] 移山所需的最少秒数
 */

// @lc code=start
public class Solution
{
    public long MinNumberOfSeconds(int mountainHeight, int[] workerTimes)
    {
        var worker = workerTimes.Length;
        long res = 0;
        var round = new int[worker];
        var times = new long[worker];
        var pq = new PriorityQueue<int, long>();

        // init pq
        for (var i = 0; i < worker; i++)
        {
            round[i] = 0;
            times[i] = 0;
            pq.Enqueue(i, workerTimes[i]);
        }

        while (mountainHeight > 0)
        {
            var key = pq.Dequeue();
            round[key]++;
            times[key] += (long)workerTimes[key] * round[key];
            pq.Enqueue(key, times[key] + workerTimes[key] * (round[key] + 1));
            mountainHeight--;
        }

        foreach (var t in times)
        {
            res = Math.Max(res, t);
        }

        return res;
    }
}
// @lc code=end

File: Answer/3296.移山所需的最少秒数.java
/*
 * @lc app=leetcode.cn id=3296 lang=java
 *
 * [3296] 移山所需的最少秒数
 */

// @lc code=start
import java.util.PriorityQueue;

class Solution {

    static class Worker {

        int index;
        long nextTime;

        Worker(int index, long nextTime) {
            this.index = index;
            this.nextTime = nextTime;
        }
    }

    public long minNumberOfSeconds(int mountainHeight, int[] workerTimes) {
        int workers = workerTimes.length;
        long res = 0;
        int[] round = new int[workers];
        long[] times = new long[workers];

        PriorityQueue<Worker> pq = new PriorityQueue<>((a, b) -> Long.compare(a.nextTime, b.nextTime));

        // init pq
        for (int i = 0; i < workers; i++) {
            round[i] = 0;
            times[i] = 0;
            pq.offer(new Worker(i, workerTimes[i]));
        }

        while (mountainHeight > 0) {
            Worker w = pq.poll();
            int key = w.index;
            round[key]++;
            times[key] += (long) workerTimes[key] * round[key];
            long nextTime = times[key] + (long) workerTimes[key] * (round[key] + 1);
            pq.offer(new Worker(key, nextTime));
            mountainHeight--;
        }

        for (long t : times) {
            res = Math.max(res, t);
        }

        return res;
    }
}
// @lc code=end

